source/Calamari.Aws/Deployment/Conventions/BucketKeyProvider.cs

```diff
@@ -1,26 +1,93 @@
+        string EncodeBucketKeyForUrl(string bucketKey);
     }
     
     public class BucketKeyProvider : IBucketKeyProvider
     {
-        public string GetBucketKey(string defaultKey, IHaveBucketKeyBehaviour behaviour)
+        public string GetBucketKey(string defaultKey, IHaveBucketKeyBehaviour behaviour, string packageFilePath = "")
         {
+            var packageContentHash = string.Empty;
+            if (!packageFilePath.IsNullOrEmpty())
+            {
+                packageContentHash = CalculateContentHash(packageFilePath);
+            }
             switch (behaviour.BucketKeyBehaviour)
             {
                 case BucketKeyBehaviourType.Custom:
-                    return behaviour.BucketKey;
+                    return SubstitutePackageHashVariable(behaviour.BucketKey, packageContentHash);
                 case BucketKeyBehaviourType.Filename:
-                    return $"{behaviour.BucketKeyPrefix}{defaultKey}";
+                    return $"{SubstitutePackageHashVariable(behaviour.BucketKeyPrefix, packageContentHash)}{defaultKey}";
+                case BucketKeyBehaviourType.FilenameWithContentHash:
+                    var (fileName, extension) = GetFileNameParts(defaultKey);
+                    var bucketKey = new StringBuilder();
+                    bucketKey.Append(SubstitutePackageHashVariable(behaviour.BucketKeyPrefix, packageContentHash));
+                    bucketKey.Append(fileName);
+                    if (!packageContentHash.IsNullOrEmpty()) bucketKey.Append($"@{packageContentHash}");
+                    bucketKey.Append(extension);
+                    return bucketKey.ToString();
                 default:
                     throw new ArgumentOutOfRangeException(nameof(behaviour), "The provided bucket key behavior was not valid.");
             }
         }
+
+        string CalculateContentHash(string packageFilePath)
+        {
+            var packageContent = File.ReadAllBytes(packageFilePath);
+            using (SHA256 sha256Hash = SHA256.Create())
+            {
+                var computedHashByte = sha256Hash.ComputeHash(packageContent);
+                var computedHash = new StringBuilder();
+                foreach (var c in computedHashByte)
+                {
+                    computedHash.Append(c.ToString("X2"));
+                }
+
+                return computedHash.ToString();
+            }
+        }
+
+        string SubstitutePackageHashVariable(string input, string contentHash)
+        {
+            return input.Replace($"#{{Octopus.Action.Package.PackageContentHash}}", contentHash);
+        }
+        
+        public string EncodeBucketKeyForUrl(string bucketKey)
+        {
+            var prefix = Path.GetDirectoryName(bucketKey)?.Replace('\\','/') ?? string.Empty;
+            var (fileName, extension) = GetFileNameParts(Path.GetFileName(bucketKey));
+            return $"{prefix}/{Uri.EscapeDataString(fileName)}{extension}";
```

Comment:
If `prefix` is empty the bucket key will be `/{filename}`, if `prefix` is a folder the bucket key will be `folder//{filename}` ?

Reply:
nom it will be `folder/{filename}`. `Path.GetDirectoryName()` return the path without the / at the end.